va/tlsalpn.go

```diff
@@ -114,27 +127,47 @@ func (va *ValidationAuthorityImpl) tryGetChallengeCert(
 	// talking to the first IPv6 address, try the first IPv4 address
 	validationRecord.AddressUsed = v4[0]
 	address := net.JoinHostPort(v4[0].String(), validationRecord.Port)
-	cert, cs, err := va.getChallengeCert(ctx, address, identifier, tlsConfig)
+	cert, cs, err := va.getChallengeCert(ctx, address, ident)
 	return cert, cs, validationRecord, err
 }
 
 func (va *ValidationAuthorityImpl) getChallengeCert(
 	ctx context.Context,
 	hostPort string,
-	identifier identifier.ACMEIdentifier,
-	config *tls.Config,
+	ident identifier.ACMEIdentifier,
 ) (*x509.Certificate, *tls.ConnectionState, error) {
-	va.log.Info(fmt.Sprintf("%s [%s] Attempting to validate for %s %s", core.ChallengeTypeTLSALPN01, identifier, hostPort, config.ServerName))
-	// We expect a self-signed challenge certificate, do not verify it here.
-	config.InsecureSkipVerify = true
+	tlsConfig := &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		NextProtos: []string{ACMETLS1Protocol},
+		// We expect a self-signed challenge certificate, do not verify it here.
+		InsecureSkipVerify: true,
+	}
+	switch ident.Type {
+	case identifier.TypeDNS:
+		tlsConfig.ServerName = ident.Value
+	case identifier.TypeIP:
+		reverseIP, err := dns.ReverseAddr(ident.Value)
+		if err != nil {
+			va.log.Infof("%s Failed to parse IP address %s.", core.ChallengeTypeTLSALPN01, ident.Value)
+			return nil, nil, fmt.Errorf("Failed to parse IP address")
+		}
+		tlsConfig.ServerName = reverseIP
+	default:
+		// This should never happen. The calling function should check the
+		// identifier type.
+		va.log.Infof("%s Unknown identifier type '%s' for %s.", core.ChallengeTypeTLSALPN01, ident.Type, ident.Value)
+		return nil, nil, fmt.Errorf("Unknown identifier type: %s", ident.Type)
+	}
```

Comment:
rather than setting tlsConfig.ServerName here, just set a local variable. Then use that to build the whole tlsConfig object _after_ this block. We like to build objects all at once when we can, rather than setting fields on them piecemeal.